internal/backends/testing/search_api.rs

```diff
@@ -51,11 +59,98 @@ impl ElementHandle {
+    ///
+    /// ```slint,no-preview
+    /// component MyView {
+    ///    submit-button := Button {}
+    /// }
+    /// export component App {
+    ///     VerticalLayout {
+    ///         close := Button {}
+    ///     }
+    /// }
+    /// ```
+    pub fn find_by_element_id(
+        component: &impl i_slint_core::api::ComponentHandle,
+        id: &str,
+    ) -> impl Iterator<Item = Self> {
+        // dirty way to get the ItemTreeRc:
+        let item_tree = WindowInner::from_pub(component.window()).component();
+        let result = search_item(&item_tree, |elem| {
+            elem.element_type_names_and_ids().unwrap().any(|(_, eid)| eid == id)
         });
-        result.into_iter().map(ElementHandle)
+        result.into_iter()
+    }
+
+    /// This function searches through the entire tree of elements of `component`, looks for
+    /// elements with given type name.
+    pub fn find_by_element_type_name(
+        component: &impl i_slint_core::api::ComponentHandle,
+        type_name: &str,
+    ) -> impl Iterator<Item = Self> {
+        // dirty way to get the ItemTreeRc:
+        let item_tree = WindowInner::from_pub(component.window()).component();
+        let result = search_item(&item_tree, |elem| {
+            elem.element_type_names_and_ids().unwrap().any(|(tn, _)| tn == type_name)
+        });
+        result.into_iter()
+    }
+
+    /// Returns an iterator over tuples of element type names and their ids. Returns None if the
+    /// element is not valid anymore.
+    ///
+    /// Elements can have multiple type names and ids, due to inheritance.
+    /// In the following example, the `PushButton` element returns for `element_type_names_and_ids`
+    /// the following tuples:
+    /// entries:
+    ///   * ("PushButton", "App::mybutton")
+    ///   * ("ButtonBase", "PushButton::root")
+    ///   * ("", "ButtonBase::root")
+    ///
+    /// ```slint,no-preview
+    /// component ButtonBase {
+    ///    // ...
+    /// }
+    /// component PushButton inherits ButtonBase {
+    /// }
+    /// export component App {
+    ///     mybutton := PushButton {}
+    /// }
+    /// ```
+    ///
+    /// ```rust
+    /// # i_slint_backend_testing::init_no_event_loop();
+    /// # slint::slint!{ #[debug_info]
+    /// # component ButtonBase { }
+    /// # component PushButton inherits ButtonBase { }
+    /// # export component App {
+    /// #    mybutton := PushButton {}
+    /// # }
+    /// # }
+    /// let app = App::new().unwrap();
+    /// let button = i_slint_backend_testing::ElementHandle::find_by_element_id(&app, "App::mybutton")
+    ///              .next().unwrap();
+    /// assert_eq!(button.element_type_names_and_ids().unwrap().collect::<Vec<_>>(),
+    ///           [("PushButton".into(), "App::mybutton".into()),
+    ///            ("ButtonBase".into(), "PushButton::root".into()),
+    ///            ("".into(), "ButtonBase::root".into())
+    ///           ]);
+    /// ```
+    pub fn element_type_names_and_ids(
```

Comment:
If it was me, I'd only give the "App::mybutton" and not the root for everything.
I'd have `element_id(&self) -> Option<SharedString>` and `element_type(&self) -> Option<SharedString>` as single function.

This means you cannot match based on the root. 
Maybe we could do `element_bases(&self) -> Option<impl Iterator<Item = SharedString>>` so we could still match on the root.

Reply:
Good idea, that makes the API easier to use for the common case.

Reply:
This is implemented now (also for C++).